Auto start on Enter key, encrypt temporary files, other small improvements

HACKERALERT opened the pull request:
Closes #106

@Retengart @davehouser1

Read the changes in Changelog.md to see exact changes of this PR.

HACKERALERT commented:
This PR probably needs more careful review as a relatively large number of changes that will affect most users are present.

HACKERALERT commented on src/Picocrypt.go:
```diff
@@ -173,10 +177,117 @@ func (p *compressorProgress) Read(data []byte) (int, error) {
+		return
+	}
+
+	if keyfile && keyfiles == nil {
```
resetUI() always sets keyfiles to nil so this is fine.

davehouser1 commented:
I tried using @Retengart's version [here ](https://github.com/Retengart/PicoDwarFS/pull/1), but pressing Enter does not do anything. I would expect that pressing Enter would run Decrypt or Encrypt button by default. Just to be specific on what I am doing. 
- Open Picocrypt.exe
- Drag and drop encrypted file into picocrypt
- Enter in password
- Press Enter
- Nothing happens.

Forgive me, I approved @HACKERALERT  prematurely without testing. Unfortunately I don't have a development env to try and compile to a `.exe`, so I cant test HACKERALERT's version. Just to confirm has the process I have above been tested?

Retengart commented:
@davehouser1 , double Enter...

davehouser1 commented:
> @davehouser1 , double Enter in my version

That worked, thanks. I can't tell from the code, is this code single enter or double enter?

HACKERALERT commented:
Give me a sec @davehouser1 , I will have .exe for you to test

HACKERALERT commented:
@davehouser1 https://github.com/Picocrypt/Picocrypt/actions/runs/14386198892?pr=108 Download the zip file and inside is a Picocrypt.exe with the changes from this PR. If I push any updates to this PR, you will need to download the updated artifacts. To do this, go to https://github.com/Picocrypt/Picocrypt/actions/workflows/pr-test-build-windows.yml, click into the newest, and download the zip file. Try my version of Enter key and let me know what you think.

Same with @Retengart, you know where to find build artifacts for your platform.

**Anyone reading, I definitely need more testing of this PR. Test rigorously and let me know of any issues.**

I'm signing off for a day or two to study for exams, so don't expect anything from me.

davehouser1 commented:
@HACKERALERT Thank you. Tested, and pressing enter works fine in Windows.
Be aware, I am not a good test pipeline. I performed the following test case:
- Opened Picocrypt.exe
- Dragged and dropped a new file into it.
- Added password, confirmed it.
- Checked Paranoid mode, Reed-Solomon, Deniability, and Delete files
- Selected Enter to Encrypt (Worked)
- Moved encrypted file back
- Put in the password
- Selected Enter, it decrypted.

One comment, might be out of scope of this PR. If the unencrypted file already exists, a warning will pop up (expected), however I still need to use the mouse to select "Yes". I am unable to select TAB to move between the two options. I suggest, having "No" be highlighted by default, in case someone hits Enter too quickly twice. Then if the user selects TAB it will switch to "Yes" and then they can hit Enter there too to confirm.

HACKERALERT commented:
Great, this is reassuring. I'll do some more testing before I merge.

On the side, @CodeCracker-oss @Retengart @davehouser1 this PR also contains warning/info for (1) encrypting multiple files and (2) encrypting to an external destination. What are your opinions? I don't want it to be too intrusive but an unencrypted temporary zip file on unencrypted external media is a big problem that I didn't catch somehow in the past. For large files, requiring double the storage space for a temporary zip file is also perhaps not intuitive. If any of you have any ideas of what an ideal handling of these two things are, I would be happy to hear. Thanks.

HACKERALERT commented:
Additional things I want to explore (for my own record):

Removing goroutines to prevent unsynchronized var access/updates (somehow this hasn't caused any issues at all)
Update zip extracting to set rw------- permissions (no x) to improve security
Check that important buffers/slices/arrays are non-zero/non-nil, panic if so
Check all errors, panic on ones that shouldn't happen

HACKERALERT commented:
Okay, so the "warning" is not clear enough. I think I should make a pop up like "Overwrite existing files?" that gives the user no way of missing it. I'll play around with this.

Ideally yeah streaming the zip file contents as it is being made in memory is the best, but I don't see a straightforward way to do so. But I'll take a look again because it would be the best solution.

HACKERALERT commented:
Okay, following Edit 4 above, everything works, the unencrypted temporary zip file is now encrypted with a random key that is forgotten when encryption is done. So no longer need to worry about unencrypted temporary files ever again. Only one question remaining, should Picocrypt/how should Picocrypt inform the user that a temporary file will be used (since it will require double free disk space)

HACKERALERT commented:
So there's generally no reliable way to tell if a filepath is on external media? Without calling shell commands that is, and only inferring from the path string itself. Do you think checking for "/media/" and "/mnt" is sufficient? Or maybe to be even more strict and say anything outside of /home is potentially an external drive.

The problem is that there's a `total` variable used through the encryption process and this `total` is the file size of the input file, which in the case of multiple file encryption, is the size of the temporary zip file. I suppose I can hack something together but it wouldn't be ideal

Possible way 1: create a temporary .zip that discards all writes, aka just simulates making a zip file to get the final file size. But this requires reading all input files twice and assumes creating a .zip has a deterministic file size
Possible way 2: maintain two branches of logic for RS and with RS; possible but not ideal.

Still need to think more about this.

HACKERALERT commented:
Thanks, I'll keep the /media/ and /mnt/ check then. Wait, actually, now that the temporary zip file is encrypted, it doesn't even matter any more, so I can just remove the external media check completely. The only thing to look out for is the double storage space requirement, which I'm still thinking how to inform the user.

Yeah the Reed-Solomon code is very slow despite having amd64 assembly that is supposed to accelerate it. I don't usually use RS since modern storage media is reliable enough and have some of their own ECC probably, but if I was to backup to a hard drive, I probably would tick it. I agree, temporary files are a bigger issue that need more thought. I think it's best to leave that as a separate topic/future task and focus on encrypting temporary files and the start on Enter key for this PR.

HACKERALERT commented:
> When a user drags in a folder or multiple files where it then performs archiving, could you then have the warning printed in red just below the area where it states the folder/file count at the top?

I'll give this a try. I'm not sure if there's enough space in the UI and may need a newline.

> Another idea is It could also perform a check of available space on local drive to ensure it will have enough space to perform the task, and if not warn the user through a pop-up, to avoid unnecessary annoyances.

I've thought about this as well but I'm not sure if there is a cross platform way or library to do these checks.

I think RS is still a valuable selling point that other tools like 7-Zip and VeraCrypt don't have, so it's worth keeping. Performance is bad though, that's for sure.

HACKERALERT commented:
For anyone in this thread, you can click into pr-test-build-{linux, macos, windows} runs above "Merging is blocked" to get test executables with this PR's changes without compiling locally.

HACKERALERT commented:
Need as many people to help me carefully test these changes in this PR:
<ul>
	<li>✓ Allow pressing 'Enter' key to press Start/Process button</li>
	<li>✓ Update "Encrypt" button to "Zip and Encrypt" if multiple files</li>
	<li>✓ Give user estimated required free disk space in status label</li>
	<li>✓ Encrypt previously unencrypted temporary zip files</li>
	<li>✓ Add `.incomplete` to filenames while work is in progress</li>
	<li>✓ Use `encrypted-*.zip.pcv` output name instead of `Encrypted.zip.pcv`</li>
	<li>✓ Use 0600 permissions when auto unzipping for optimal security</li>
</ul>

There's a lot of them and basically anyone who uses Picocrypt will be running updated code, so it's very important to test thoroughly before merging. Please let me know of any issues, and if none, would still like to get a comment so I know.

HACKERALERT commented:
Not sure, but .tmp sort of implies it is a temporary file that can be deleted already. although wouldn't hurt to add a .incomplete to the end of that, but there's a lot of inner processes and it might be fine just to have the most important thing (the output) suffixed with .incomplete

HACKERALERT commented:
Change 0600 to 0755, does that fix it?

HACKERALERT commented:
Hmm the problem with 755 before is it allows executing extracted files which probably isn't a secure default. Can you try 0700, 0666, 0660?

HACKERALERT commented:
Oh lol I was doing 0600 on directories 😂

Try 0700, that would be the ideal security wise.

HACKERALERT commented:
Great, thanks, I will update that.

HACKERALERT commented:
I'm gonna be off for a few days. Let me know of any testing results.

HACKERALERT commented:
Tab is supposed to be handled by the UI library, so I would have to look more deeply. In this PR, volume names for multiple-file volumes are now `encrypted-unixtimestamp.whatever.pcv` so you should not encounter overwriting. Unless you are decrypting a single file, but why would you decrypt a file when it's already there unencrypted? I don't think most people will encounter the overwrite confirmation so I think it's fine to require manual mouse input.

davehouser1 commented:
Fair point. The situation arises for me because the file I'm encrypting gets auto generated if it does not exist. This is an env var file for an application. It will generate a new empty file if it can't find the file. I'v run into situations where I open the application by accident before unencrypting the env var file, so the program creates a new one. I then go to unencrypt the file, but the default auto created one exists.
Explaining for context. However, I agree, it's and edge case. I'm fine with just using the mouse for that operation. Glad the enter key works for encrypt and decrypt in this PR :)

HACKERALERT commented:
Interesting case, thanks for the additional info. Yeah I think I'll keep it as is, at least for this PR.

HACKERALERT commented:
Any final things to say before I do a few minor changes and merge?

HACKERALERT commented:
I'm going to update the storage size estimator to be a little better and I'll merge by EOD today.

HACKERALERT commented:
Okay in the interest of stability, I will be doing a full code review and trying to add error handlings in *all* places where an error is thrown. Current code discards a lot of errors. This will take some time, so not merging today.

In the meanwhile, if any of you are free, help me test the "ensure X free space is available" that I've just updated. It should now be very accurate and account for deniability and split volumes. Single file encryption is 1x the input file size, multiple files encrypt requires a temporary .zip so the multiplier is 2x, if deniability is enabled, make it 3x since we'll need to encrypt again, and if split/recombine on encrypt/decrypt is needed, increment multiplier again since we'll need to split or recombine. Try different combinations. The label should correctly show the exact amount of free space that will be used.

HACKERALERT commented:
@davehouser1 @Retengart @CodeCracker-oss 

Sorry for the ping, hopefully it will be the final for this PR. I did a lot of code additions and now *all* errors everywhere in the code are checked for. Great addition for sure, but added a lot of code and cases. Can I get a final sanity check that nothing broke? Basically every flow in Picocrypt's code has been changed so everything needs to be tested, every feature and combination of features 💀 

...only if you are bored and have some free time. Otherwise I'll merge today or tomorrow or maybe the day after that.

HACKERALERT commented:
read the code 🤦

HACKERALERT commented:
Great to hear, thanks for the help!

What is it? You can tell me here and if it's a bigger issue then you can open a new one

HACKERALERT commented:
Actually open a new issue since it's present in 47

HACKERALERT commented:
Thanks for all the help! Here goes.
